extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,165 @@
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullurl_ = this.url_;
+
+    /** @private {string} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.timeout_ = AD_BATCH_TIMEOUT;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Get a list of slots that are in use for this URL
+    const slots = [];
+    const elements = element.win.imageadElements[this.url_];
+    for (let i = 0; i < elements.length; i++) {
+      const e = elements[i];
+
+      // If a slot is defined, add it to the list if it isn't there already
+      let s = e.getAttribute('data-slot');
+      if (s === null) {
+        s = 0;
+      }
+      let done = false;
+      for (const x in slots) {
+        if (slots[x] == s) {
+          done = true;
+        }
+      }
+      if (!done) {
+        slots.push(s);
+      }
+    }
+    // If any slots were defined, add them to the URL
+    if (slots.length) {
+      this.fullurl_ += '?s=' + slots.join(',');
+    }
+  }
+
+  /**
+   * Call this to instruct the batch manager to do the layout for an element. Prior to doing the layout, the batch manager
+   * will fetch the data (if this is the batch master) or wait for it to be ready (if not the batch master).
+   * @param {Element} e The element for which the layout is to be done
+   * @returns {Promise} The element's "jsondata_" variable will be filled its with the data from the ad server before this is
+   *     resolved.
+   */
+  doLayout(e) {
+    const t = this;
+    return new Promise(function(resolve) {
+      if (e.isBatchMaster_) {
+        // This is the master. We're going to do the layout for all ads in the batch
+        // Gather the parameters to be added to the URL. First, make a list of the slot ids for all ads on this page with this URL
+        xhrFor(e.win).fetchJson(t.fullurl_).then(function(data) {
+          t.responseData = data;
+          resolve('OK');
+        }, function(err) {
+          user().error(TAG_AD_BATCH_MANAGER, err);
+        });
+
+      } else {
+        // This is not the master. We need to wait for the master to have fetched the response for the ad server; when this is done,
+        // it's OK to proceed
+        t.getResponseWhenReady().then(function() {
+          resolve('OK');
+        }, function(err) {
+          user().error(TAG_AD_BATCH_MANAGER, err);
+        });
+      }
+    });
+  }
+
+  /**
+   * Return a promise that will complete when the response from the image server has been received.
+   * If the response is not ready after the specified timeout, reject the promise.
+   * @returns {Promise}
+   */
+  getResponseWhenReady() {
+    const t = this;
+    return new Promise(function(resolve, reject) {
+      if (t.responseData === null) {
+        // The response data is not ready. Let's wait 50 ms and try again
+        if (--t.timeout_) {
```

Comment:
It's needed in case the doLayout() gets called in the wrong order - i.e. a slave gets called before the master (which you told me was a possibility). There may be other ways of doing it (e.g. your responsePromise suggestion above), but I think what's here is easy to understand.